src/Network/Minio/S3API.hs

```diff
@@ -358,3 +366,31 @@ headBucket bucket = headBucketEx `catches`
                                    , riBucket = Just bucket
                                    }
       return $ NC.responseStatus resp == HT.ok200
+
+makePresignedURL :: UrlExpiry -> HT.Method -> Maybe Bucket -> Maybe Object
+                 -> Region -> HT.Query -> HT.RequestHeaders -> Minio ByteString
+makePresignedURL expiry method bucket object region queryParams headers = do
+  ci <- asks mcConnInfo
+
+  let
+    hostHeader = (hHost, formatBS "{}:{}" [connectHost ci,
+                                           show $ connectPort ci])
+    ri = def { riMethod = method
+             , riBucket = bucket
+             , riObject = object
+             , riQueryParams = queryParams
+             , riHeaders = hostHeader : headers
+             , riRegion = Just region
+             }
+
+  signPairs <- liftIO $ signV4 ci ri (Just expiry)
+
+  let
+    qpToAdd = map (\(x, y) -> (x, Just y)) signPairs
+    proto = bool "http://" "https://" $ connectIsSecure ci
+
+  return $ B.concat [ proto
+                    , formatBS "{}:{}" (connectHost ci, connectPort ci)
```

Comment:
we could include all the parts of presignURL as parameters for `formatBS`. This would make the URL structure obvious in code. E.g,
`formatBS "{}://{}:{}/{}{}" (host, port, resource, queryParams)`

Reply:
format params need to have show instances.